sdcm/utils/nemesis_utils/node_allocator.py

```diff
@@ -0,0 +1,188 @@
+
+            candidate_nodes = self._filter_nodes(
+                available_for_selection, is_seed=is_seed, dc_idx=dc_idx, rack=rack, filter_seed=filter_seed)
+
+            if not candidate_nodes:
+                dc_str = f'dc_idx={dc_idx}' if dc_idx is not None else ''
+                rack_str = f'rack={rack}' if rack is not None else ''
+                seed_str = f'is_seed={is_seed}' if is_seed is not DefaultValue else ''
+                filter_str = ", ".join([dc_str, rack_str, seed_str])
+                reason = (
+                    f"The following pool '{pool_type.value}' nodes are available (not busy) for selection, but none of "
+                    f"them match the specified criteria ({filter_str}).\n"
+                    f"{[n.name for n in available_for_selection]}.")
+                LOGGER.warning("%s: node allocation failed. %s", nemesis_name, reason)
+                return None, reason
+
+            selected_node: BaseNode
+            if allow_only_last_node_in_rack and rack is not None:
+                selected_node = candidate_nodes[-1]
+            else:
+                selected_node = random.choice(candidate_nodes)
+
+            self.active_nemeses_on_nodes[selected_node] = nemesis_name
+            LOGGER.info("%s: selected target node %s. Marked it as busy by allocator.",
+                        nemesis_name, selected_node.name)
+            return selected_node, None
+
+    def mark_node_as_busy(self, node: BaseNode, nemesis_name: str):
+        """
+        Marks a node as busy with a nemesis.
+
+        This is for the cases when e.g. a new node is added by a nemesis, which is running on another target node.
+        """
+        with self.lock:
+            if node in self.active_nemeses_on_nodes:
+                LOGGER.warning("%s: Tried to mark node %s as busy, but it was already busy with %s.",
+                               nemesis_name, node.name, self.active_nemeses_on_nodes[node])
+                return False
+
+            self.active_nemeses_on_nodes[node] = nemesis_name
+            LOGGER.info("%s: marked node %s as busy.", nemesis_name, node.name)
+            return True
+
+    def release_busy_node(self, node: BaseNode, nemesis_name: str):
+        """Releases the node, making it available for other nemeses."""
+        with self.lock:
+            if node in self.active_nemeses_on_nodes:
+                if self.active_nemeses_on_nodes[node] == nemesis_name:
+                    del self.active_nemeses_on_nodes[node]
+                    LOGGER.info("%s: released busy node %s.", nemesis_name, node.name)
+                else:
+                    LOGGER.warning("%s: tried to release node %s, but it was busy with a different nemesis %s.",
+                                   nemesis_name, node.name, self.active_nemeses_on_nodes[node])
+            else:
+                LOGGER.warning("%s: tried to release node %s, but it was not marked as busy by the allocator.",
```

Comment:
Just a note: We should monitor these closely, as they most likely indicate implementation error

Reply:
Hence we should raise as an error event, warning wouldn't be noticed.